File: matrixos/emoji_loader.py
"""
Emoji Sprite Sheet Loader for MatrixOS

Loads and extracts emojis from the bundled sprite sheet for efficient
offline icon support. Can also download emojis on-demand if enabled.
"""

import os
import json
import urllib.request
import io
from pathlib import Path
from PIL import Image


class EmojiLoader:
    """Loads emojis from sprite sheet."""
    
    def __init__(self, spritesheet_path=None, metadata_path=None):
        """Initialize emoji loader.
        
        Args:
            spritesheet_path: Path to sprite sheet PNG (default: matrixos/emoji_spritesheet.png)
            metadata_path: Path to metadata JSON (default: matrixos/emoji_spritesheet.json)
        """
        # Default paths relative to this file
        if spritesheet_path is None:
            base_dir = os.path.dirname(os.path.abspath(__file__))
            spritesheet_path = os.path.join(base_dir, 'emoji_spritesheet.png')
        
        if metadata_path is None:
            base_dir = os.path.dirname(os.path.abspath(__file__))
            metadata_path = os.path.join(base_dir, 'emoji_spritesheet.json')
        
        self.spritesheet_path = spritesheet_path
        self.metadata_path = metadata_path
        
        # Lazy load
        self._spritesheet = None
        self._metadata = None
        self._cache = {}  # Cache extracted emoji images
    
    def _load_spritesheet(self):
        """Load sprite sheet image (lazy)."""
        if self._spritesheet is None:
            if not os.path.exists(self.spritesheet_path):
                raise FileNotFoundError(f"Sprite sheet not found: {self.spritesheet_path}")
            self._spritesheet = Image.open(self.spritesheet_path)
        return self._spritesheet
    
    def _load_metadata(self):
        """Load metadata JSON (lazy)."""
        if self._metadata is None:
            if not os.path.exists(self.metadata_path):
                raise FileNotFoundError(f"Metadata not found: {self.metadata_path}")
            with open(self.metadata_path, 'r') as f:
                self._metadata = json.load(f)
        return self._metadata
    
    def has_emoji(self, emoji):
        """Check if emoji is available in sprite sheet.
        
        Args:
            emoji: Emoji character (e.g., "🕹️")
            
        Returns:
            True if emoji is in sprite sheet
        """
        metadata = self._load_metadata()
        return emoji in metadata['emojis']
    
    def get_emoji_image(self, emoji):
        """Extract emoji image from sprite sheet.
        
        Args:
            emoji: Emoji character (e.g., "🕹️")
            
        Returns:
            PIL Image object (32×32 RGBA), or None if not found
        """
        # Check cache first
        if emoji in self._cache:
            return self._cache[emoji]
        
        # Load metadata
        metadata = self._load_metadata()
        if emoji not in metadata['emojis']:
            return None
        
        # Get position
        emoji_data = metadata['emojis'][emoji]
        x = emoji_data['x']
        y = emoji_data['y']
        size = metadata['emoji_size']
        
        # Extract from sprite sheet
        spritesheet = self._load_spritesheet()
        emoji_img = spritesheet.crop((x, y, x + size, y + size))
        
        # Cache it
        self._cache[emoji] = emoji_img
        
        return emoji_img
    
    def emoji_to_icon_json(self, emoji, size=32):
        """Convert emoji to MatrixOS icon JSON format.
        
        Args:
            emoji: Emoji character (e.g., "🕹️")
            size: Target size (will resize if needed)
            
        Returns:
            Icon data dict with 'width', 'height', 'data' keys, or None if not found
        """
        img = self.get_emoji_image(emoji)
        if img is None:
            return None
        
        # Resize if needed
        if size != 32:
            img = img.resize((size, size), Image.Resampling.LANCZOS)
        
        # Convert to RGB565 format for MatrixOS
        pixels = []
        for y in range(size):
            for x in range(size):
                r, g, b, a = img.getpixel((x, y))
                
                # Skip transparent pixels
                if a < 128:
                    continue
                
                # Skip black pixels (background/anti-aliasing artifacts)
                if r < 30 and g < 30 and b < 30:
                    continue
                
                # Convert to RGB565
                r5 = (r >> 3) & 0x1F
                g6 = (g >> 2) & 0x3F
                b5 = (b >> 3) & 0x1F
                rgb565 = (r5 << 11) | (g6 << 5) | b5
                
                pixels.append({
                    'x': x,
                    'y': y,
                    'c': rgb565
                })
        
        return {
            'width': size,
            'height': size,
            'data': pixels
        }
    
    def list_available_emojis(self, limit=None):
        """List all available emojis.
        
        Args:
            limit: Maximum number to return (None = all)
            
        Returns:
            List of emoji characters
        """
        metadata = self._load_metadata()
        emojis = list(metadata['emojis'].keys())
        
        if limit:
            return emojis[:limit]
        return emojis
    
    def download_emoji_on_demand(self, emoji, size=32):
        """Download emoji from Noto Emoji GitHub on-demand.
        
        Args:
            emoji: Emoji character to download
            size: Target size (default 32)
            
        Returns:
            PIL Image object or None if download failed
        """
        # Convert emoji to codepoint
        codepoint = self._emoji_to_codepoint(emoji)
        if not codepoint:
            return None
        
        # Download from Noto Emoji repository
        url = f"https://raw.githubusercontent.com/googlefonts/noto-emoji/main/png/{size}/emoji_u{codepoint}.png"
        
        try:
            with urllib.request.urlopen(url, timeout=5) as response:
                img_data = response.read()
                img = Image.open(io.BytesIO(img_data))
                
                # Cache it
                self._cache[emoji] = img
                
                # Optionally save to cache directory
                self._save_to_cache(emoji, img, size)
                
                return img
        except Exception as e:
            print(f"Failed to download emoji '{emoji}': {e}")
            return None
    
    def _emoji_to_codepoint(self, emoji):
        """Convert emoji character to codepoint string.
        
        Args:
            emoji: Emoji character
            
        Returns:
            Codepoint string like "1f579" or None
        """
        try:
            # Handle multi-codepoint emojis (like flags, skin tones, etc.)
            codepoints = [f"{ord(c):04x}" for c in emoji]
            return "_".join(codepoints)
        except:
            return None
    
    def _save_to_cache(self, emoji, img, size):
        """Save downloaded emoji to cache directory.
        
        Args:
            emoji: Emoji character
            img: PIL Image object
            size: Image size
        """
        try:
            # Get cache directory from system config
            try:
                from matrixos.system_config_loader import get_emoji_cache_dir
                cache_dir = get_emoji_cache_dir()
            except:
                # Fallback if system config not available
                cache_dir = Path.home() / ".matrixos" / "emoji_cache"
            
            cache_dir.mkdir(parents=True, exist_ok=True)
            
            codepoint = self._emoji_to_codepoint(emoji)
            cache_file = cache_dir / f"emoji_{codepoint}_{size}.png"
            
            img.save(cache_file, "PNG")
        except Exception as e:
            # Silent fail - caching is optional
            pass
    
    def get_emoji_with_fallback(self, emoji, size=32, allow_download=True):
        """Get emoji image with fallback to download if not in sprite sheet.
        
        Args:
            emoji: Emoji character
            size: Target size
            allow_download: Whether to download if not found (checks system config)
            
        Returns:
            PIL Image object or None
        """
        # Try sprite sheet first
        img = self.get_emoji_image(emoji)
        if img and size != 32:
            img = img.resize((size, size), Image.Resampling.LANCZOS)
        
        if img:
            return img
        
        # Not in sprite sheet - check if downloads enabled
        if not allow_download:
            return None
        
        try:
            from matrixos.system_config_loader import is_emoji_download_enabled
            if not is_emoji_download_enabled():
                return None
        except:
            # If system config not available, don't download
            return None
        
        # Try to download
        return self.download_emoji_on_demand(emoji, size)
    
    def get_info(self):
        """Get sprite sheet info.
        
        Returns:
            Dict with version, size, total_emojis, etc.
        """
        metadata = self._load_metadata()
        return {
            'version': metadata.get('version', 'unknown'),
            'emoji_size': metadata.get('emoji_size', 32),
            'columns': metadata.get('columns', 0),
            'rows': metadata.get('rows', 0),
            'total_emojis': metadata.get('total_emojis', 0),
            'available_emojis': len(metadata.get('emojis', {}))
        }


# Global singleton instance
_loader = None

def get_emoji_loader():
    """Get global emoji loader instance."""
    global _loader
    if _loader is None:
        _loader = EmojiLoader()
    return _loader


if __name__ == '__main__':
    # Test the loader
    loader = EmojiLoader()
    
    print("="*60)
    print("Emoji Sprite Sheet Loader Test")
    print("="*60)
    
    info = loader.get_info()
    print(f"Version: {info['version']}")
    print(f"Emoji size: {info['emoji_size']}×{info['emoji_size']}")
    print(f"Grid: {info['columns']}×{info['rows']}")
    print(f"Total emojis: {info['available_emojis']}")
    
    # Test a few emojis
    test_emojis = ['🕹️', '🎮', '👾', '🐍', '🧱', '🏓', '😂', '❤️', '☀️', '⏰']
    
    print("\nTesting emoji extraction:")
    for emoji in test_emojis:
        if loader.has_emoji(emoji):
            img = loader.get_emoji_image(emoji)
            icon = loader.emoji_to_icon_json(emoji)
            print(f"  ✅ {emoji} - {img.size if img else 'N/A'} - {len(icon['data']) if icon else 0} pixels")
        else:
            print(f"  ❌ {emoji} - not found")
    
    print("\nFirst 20 emojis:")
    first_20 = loader.list_available_emojis(limit=20)
    print("  " + " ".join(first_20))
    
    print("\n✨ Emoji loader working!")

File: pizxel/lib/emoji-loader.ts
/**
 * Emoji Sprite Sheet Loader for PiZXel
 *
 * Loads and extracts emojis from the bundled sprite sheet for efficient
 * offline icon support.
 */

import { createCanvas, loadImage, Image as CanvasImage } from "canvas";
import * as fs from "fs";
import * as path from "path";
import * as https from "https";
import { isNetworkAllowed } from "../core/network";
import { DisplayBuffer } from "../core/display-buffer";

interface EmojiMetadata {
  version: string;
  emoji_size: number;
  columns: number;
  rows: number;
  total_emojis: number;
  emojis: {
    [emoji: string]: {
      codepoint: string;
      row: number;
      col: number;
      x: number;
      y: number;
    };
  };
}

interface EmojiImageData {
  width: number;
  height: number;
  data: Uint8ClampedArray; // RGBA pixel data
}

export class EmojiLoader {
  private spritesheetPath: string;
  private metadataPath: string;
  private spritesheet: CanvasImage | null = null;
  private metadata: EmojiMetadata | null = null;
  private cache: Map<string, EmojiImageData> = new Map();

  constructor(spritesheetPath?: string, metadataPath?: string) {
    // Default paths relative to this file
    const baseDir = __dirname;

    this.spritesheetPath =
      spritesheetPath || path.join(baseDir, "emoji_spritesheet.png");
    this.metadataPath =
      metadataPath || path.join(baseDir, "emoji_spritesheet.json");
  }

  /**
   * Load sprite sheet image (lazy)
   */
  private async loadSpritesheet(): Promise<CanvasImage> {
    if (this.spritesheet === null) {
      if (!fs.existsSync(this.spritesheetPath)) {
        throw new Error(`Sprite sheet not found: ${this.spritesheetPath}`);
      }
      this.spritesheet = await loadImage(this.spritesheetPath);
    }
    return this.spritesheet;
  }

  /**
   * Load metadata JSON (lazy)
   */
  private loadMetadata(): EmojiMetadata {
    if (this.metadata === null) {
      if (!fs.existsSync(this.metadataPath)) {
        throw new Error(`Metadata not found: ${this.metadataPath}`);
      }
      const data = fs.readFileSync(this.metadataPath, "utf-8");
      this.metadata = JSON.parse(data);
    }
    return this.metadata!;
  }

  /**
   * Get codepoint string for Twemoji CDN URL
   */
  private getEmojiCodepoint(emoji: string): string {
    const codePoints: string[] = [];
    for (let i = 0; i < emoji.length; i++) {
      const code = emoji.codePointAt(i);
      if (code !== undefined) {
        // Skip variation selectors (U+FE00 to U+FE0F)
        if (code >= 0xfe00 && code <= 0xfe0f) {
          if (code > 0xffff) i++; // Skip low surrogate
          continue;
        }
        codePoints.push(code.toString(16));
        // Skip low surrogate if this was a high surrogate
        if (code > 0xffff) i++;
      }
    }
    return codePoints.join("-");
  }

  /**
   * Fetch emoji image from Twemoji CDN
   */
  private async fetchEmojiFromCDN(emoji: string): Promise<CanvasImage | null> {
    if (!isNetworkAllowed()) {
      return null;
    }

    try {
      const codepoint = this.getEmojiCodepoint(emoji);
      // Fetch 72×72 PNG from CDN (we'll scale it to 32×32 for caching)
      const url = `https://cdn.jsdelivr.net/gh/jdecked/twemoji@latest/assets/72x72/${codepoint}.png`;

      console.log(`[EmojiLoader] Fetching ${emoji} from CDN: ${url}`);

      return await new Promise<CanvasImage>((resolve, reject) => {
        https
          .get(url, (response) => {
            if (response.statusCode !== 200) {
              reject(new Error(`CDN returned ${response.statusCode}`));
              return;
            }

            const chunks: Buffer[] = [];
            response.on("data", (chunk) => chunks.push(chunk));
            response.on("end", async () => {
              try {
                const buffer = Buffer.concat(chunks);
                const image = await loadImage(buffer);
                resolve(image);
              } catch (err) {
                reject(err);
              }
            });
          })
          .on("error", reject);
      });
    } catch (err) {
      console.warn(`[EmojiLoader] Failed to fetch ${emoji} from CDN:`, err);
      return null;
    }
  }

  /**
   * Dynamically render an emoji using canvas (for emojis not in spritesheet)
   */
  private async renderEmojiDynamically(
    emoji: string
  ): Promise<EmojiImageData | null> {
    try {
      const size = 32; // Render at 32px to match spritesheet size

      // Try fetching from Twemoji CDN first (proper color emoji)
      const cdnImage = await this.fetchEmojiFromCDN(emoji);

      if (cdnImage) {
        // Successfully got color emoji from CDN
        const canvas = createCanvas(size, size);
        const ctx = canvas.getContext("2d");
        ctx.clearRect(0, 0, size, size);

        // Draw the emoji image, scaling to fit if needed
        const scale = Math.min(size / cdnImage.width, size / cdnImage.height);
        const w = cdnImage.width * scale;
        const h = cdnImage.height * scale;
        const x = (size - w) / 2;
        const y = (size - h) / 2;
        ctx.drawImage(cdnImage, x, y, w, h);

        const imageData = ctx.getImageData(0, 0, size, size);
        const emojiImageData: EmojiImageData = {
          width: size,
          height: size,
          data: imageData.data,
        };

        this.cache.set(emoji, emojiImageData);
        console.log(`[EmojiLoader] Successfully rendered ${emoji} from CDN`);
        return emojiImageData;
      }

      // Fallback: Render with canvas text (grayscale only)
      console.log(
        `[EmojiLoader] CDN failed, falling back to canvas rendering for ${emoji}`
      );
      const canvas = createCanvas(size, size);
      const ctx = canvas.getContext("2d");

      // Clear canvas
      ctx.clearRect(0, 0, size, size);

      // Set font and render emoji
      ctx.font = `${size}px Arial, "Apple Color Emoji", "Segoe UI Emoji", "Noto Color Emoji"`;
      ctx.textAlign = "center";
      ctx.textBaseline = "middle";
      ctx.fillText(emoji, size / 2, size / 2);

      // Get image data
      const imageData = ctx.getImageData(0, 0, size, size);

      // Node.js canvas renders emoji as grayscale - colorize based on emoji
      // Newspaper emoji: white/beige paper with black text
      const colorMap: Record<
        string,
        { light: [number, number, number]; dark: [number, number, number] }
      > = {
        "🗞️": { light: [240, 235, 220], dark: [20, 20, 20] }, // Beige paper, black text
        "🗞": { light: [240, 235, 220], dark: [20, 20, 20] }, // Alt encoding
      };

      const colors = colorMap[emoji] || {
        light: [200, 200, 200],
        dark: [50, 50, 50],
      };

      // Convert grayscale to colored version
      for (let i = 0; i < imageData.data.length; i += 4) {
        const r = imageData.data[i];
        const g = imageData.data[i + 1];
        const b = imageData.data[i + 2];
        const a = imageData.data[i + 3];

        // Skip transparent pixels
        if (a < 10) continue;

        // Grayscale value (all RGB should be same from canvas)
        const gray = r;

        // Map dark grays to dark color, light grays to light color
        if (gray < 128) {
          // Dark pixel - use dark color
          const intensity = gray / 128;
          imageData.data[i] = colors.dark[0] * intensity;
          imageData.data[i + 1] = colors.dark[1] * intensity;
          imageData.data[i + 2] = colors.dark[2] * intensity;
        } else {
          // Light pixel - use light color
          const intensity = (gray - 128) / 127;
          imageData.data[i] =
            colors.light[0] * intensity + colors.dark[0] * (1 - intensity);
          imageData.data[i + 1] =
            colors.light[1] * intensity + colors.dark[1] * (1 - intensity);
          imageData.data[i + 2] =
            colors.light[2] * intensity + colors.dark[2] * (1 - intensity);
        }
      }

      // Cache and return
      const emojiImageData: EmojiImageData = {
        width: size,
        height: size,
        data: imageData.data,
      };

      this.cache.set(emoji, emojiImageData);
      console.log(
        `[EmojiLoader] Successfully rendered ${emoji} dynamically with colorization`
      );
      return emojiImageData;
    } catch (err) {
      console.error(
        `[EmojiLoader] Failed to render ${emoji} dynamically:`,
        err
      );
      return null;
    }
  }

  /**
   * Check if emoji is available in sprite sheet
   */
  hasEmoji(emoji: string): boolean {
    const metadata = this.loadMetadata();
    return emoji in metadata.emojis;
  }

  /**
   * Extract emoji image from sprite sheet
   */
  async getEmojiImage(emoji: string): Promise<EmojiImageData | null> {
    // Check cache first
    if (this.cache.has(emoji)) {
      return this.cache.get(emoji)!;
    }

    // Load metadata
    const metadata = this.loadMetadata();
    if (!(emoji in metadata.emojis)) {
      console.warn(
        `[EmojiLoader] Emoji ${emoji} not found in metadata, attempting to render dynamically...`
      );
      // Try dynamic rendering for missing emojis
      return await this.renderEmojiDynamically(emoji);
    }

    // Get position
    const emojiData = metadata.emojis[emoji];
    const x = emojiData.x;
    const y = emojiData.y;
    const size = metadata.emoji_size;

    // Load spritesheet
    const spritesheet = await this.loadSpritesheet();

    // Extract emoji using canvas
    const canvas = createCanvas(size, size);
    const ctx = canvas.getContext("2d");

    // Draw the emoji portion from spritesheet
    ctx.drawImage(
      spritesheet,
      x,
      y,
      size,
      size, // Source rect
      0,
      0,
      size,
      size // Dest rect
    );

    // Get pixel data
    const imageData = ctx.getImageData(0, 0, size, size);

    const result: EmojiImageData = {
      width: size,
      height: size,
      data: imageData.data,
    };

    // Cache it
    this.cache.set(emoji, result);

    return result;
  }

  /**
   * Render emoji to display buffer at specified position (SYNC - uses cache only)
   * Returns true if emoji was rendered, false if not in cache
   */
  renderToBufferSync(
    emoji: string,
    buffer: DisplayBuffer,
    x: number,
    y: number,
    targetSize?: number
  ): boolean {
    // Only render if in cache
    if (!this.cache.has(emoji)) {
      return false;
    }

    const emojiImage = this.cache.get(emoji)!;
    const size = targetSize || emojiImage.width;
    const scale = size / emojiImage.width;

    // Render emoji pixels to buffer
    for (let py = 0; py < emojiImage.height; py++) {
      for (let px = 0; px < emojiImage.width; px++) {
        const idx = (py * emojiImage.width + px) * 4;
        const r = emojiImage.data[idx];
        const g = emojiImage.data[idx + 1];
        const b = emojiImage.data[idx + 2];
        const a = emojiImage.data[idx + 3];

        // Skip transparent pixels
        if (a < 128) {
          continue;
        }

        // Don't skip dark pixels - they're part of the emoji content!
        // (The newspaper emoji 🗞️ has black text that needs to render)

        // Calculate scaled position
        if (scale === 1) {
          // No scaling
          buffer.setPixel(x + px, y + py, [r, g, b]);
        } else {
          // Simple nearest-neighbor scaling
          for (let sy = 0; sy < scale; sy++) {
            for (let sx = 0; sx < scale; sx++) {
              const targetX = x + Math.floor(px * scale) + sx;
              const targetY = y + Math.floor(py * scale) + sy;
              buffer.setPixel(targetX, targetY, [r, g, b]);
            }
          }
        }
      }
    }

    return true;
  }

  /**
   * Render emoji to display buffer at specified position (ASYNC)
   */
  async renderToBuffer(
    emoji: string,
    buffer: DisplayBuffer,
    x: number,
    y: number,
    targetSize?: number
  ): Promise<boolean> {
    const emojiImage = await this.getEmojiImage(emoji);
    if (!emojiImage) {
      return false;
    }

    const size = targetSize || emojiImage.width;
    const scale = size / emojiImage.width;

    // Render emoji pixels to buffer
    for (let py = 0; py < emojiImage.height; py++) {
      for (let px = 0; px < emojiImage.width; px++) {
        const idx = (py * emojiImage.width + px) * 4;
        const r = emojiImage.data[idx];
        const g = emojiImage.data[idx + 1];
        const b = emojiImage.data[idx + 2];
        const a = emojiImage.data[idx + 3];

        // Skip transparent pixels
        if (a < 128) {
          continue;
        }

        // Don't skip dark pixels - they're part of the emoji content!
        // (The newspaper emoji 🗞️ has black text that needs to render)

        // Calculate scaled position
        if (scale === 1) {
          // No scaling
          buffer.setPixel(x + px, y + py, [r, g, b]);
        } else {
          // Simple nearest-neighbor scaling
          for (let sy = 0; sy < scale; sy++) {
            for (let sx = 0; sx < scale; sx++) {
              const targetX = x + Math.floor(px * scale) + sx;
              const targetY = y + Math.floor(py * scale) + sy;
              buffer.setPixel(targetX, targetY, [r, g, b]);
            }
          }
        }
      }
    }

    return true;
  }

  /**
   * List all available emojis
   */
  listAvailableEmojis(limit?: number): string[] {
    const metadata = this.loadMetadata();
    const emojis = Object.keys(metadata.emojis);

    if (limit) {
      return emojis.slice(0, limit);
    }
    return emojis;
  }

  /**
   * Get total number of emojis available
   */
  getEmojiCount(): number {
    const metadata = this.loadMetadata();
    return metadata.total_emojis;
  }

  /**
   * Get metadata (public accessor)
   */
  getMetadata(): EmojiMetadata {
    return this.loadMetadata();
  }
}

// Singleton instance
let _emojiLoader: EmojiLoader | null = null;

/**
 * Get shared emoji loader instance
 */
export function getEmojiLoader(): EmojiLoader {
  if (_emojiLoader === null) {
    _emojiLoader = new EmojiLoader();
  }
  return _emojiLoader;
}
